TypeCobol/Compiler/Scopes/Domain.cs

```diff
@@ -0,0 +1,245 @@
+﻿using System;
+using System.Collections;
+using System.Collections.Generic;
+using JetBrains.Annotations;
+using TypeCobol.Compiler.Symbols;
+
+namespace TypeCobol.Compiler.Scopes
+{
+    /// <summary>
+    /// A domain is a set of symbols.
+    /// </summary>
+    /// <typeparam name="TSymbol">Type of symbols stored in this domain.</typeparam>
+    public class Domain<TSymbol> : IEnumerable<TSymbol>
+        where TSymbol : Symbol
+    {
+        /// <summary>
+        /// Represents an entry in a domain. Regroups some symbols which have the same name.
+        /// </summary>
+        public class Entry : IEnumerable<TSymbol>
+        {
+            /*
+             * Basic optimization, an entry is initially using a single field to store its main symbol.
+             * As soon as a second symbol is added to it, we switch to a List implementation.
+             */
+
+            public string Name { get; }
+            private TSymbol _symbol;
+            private List<TSymbol> _symbols;
+
+            /// <summary>
+            /// Creates an empty Entry bound to a specific symbol name.
+            /// </summary>
+            /// <param name="name">Expected name of all symbols in this entry.</param>
+            public Entry([NotNull] string name)
+            {
+                System.Diagnostics.Debug.Assert(name != null);
+                Name = name;
+                _symbol = null;
+                _symbols = null;
+            }
+
+            /// <summary>
+            /// Creates a unique Entry.
+            /// </summary>
+            /// <param name="symbol">Symbol associated to this entry.</param>
+            public Entry([NotNull] TSymbol symbol)
+            {
+                System.Diagnostics.Debug.Assert(symbol != null);
+                System.Diagnostics.Debug.Assert(symbol.Name != null);
+                Name = symbol.Name;
+                _symbol = symbol;
+                _symbols = null;
+            }
+
+            /// <summary>
+            /// Total count of symbols in this entry.
+            /// </summary>
+            public int Count => _symbols != null ? _symbols.Count : (_symbol != null ? 1 : 0);
+
+            /// <summary>
+            /// Convenience property to retrieve the first symbol of this entry.
+            /// Returns null if this entry is empty.
+            /// </summary>
+            public TSymbol Symbol => _symbols != null ? (_symbols.Count > 0 ? _symbols[0] : null) : _symbol;
```

Comment:
Can be simplified to:
_symbols?.Count > 0 ? _symbols[0] : _symbol;

Reply:
I have been thinking about it but I think it is less readable. I can change it if you prefer.
Also here is what it would look like for Count property :
```csharp
public int Count => _symbols?.Count ?? (_symbol != null ? 1 : 0);
```

Reply:
Ok the Count

Reply:
Done for both `Count` and `Symbol`. See c00b3b2691a7b3942b0ce36ba6bfa52d3f486501.